### DISCUSSION TITLE
[New3D] JointState support

### DISCUSSION THREAD

**Discussion by jhurliman** [foxglove/community#454] [Q&A]
The end result should be the ability to load a URDF in the New3D panel and manipulate various parts of the robot using an incoming stream of JointState messages.

**Comment by amacneil** [comment]
Related feature request
* https://github.com/foxglove/community/issues/363

**Comment by amacneil** [comment]
Adding some more information:

In ROS, `JointState` messages are typically converted into `TFMessage` by [robot_state_publisher](http://wiki.ros.org/robot_state_publisher).

Rendering `JointState` without `TFMessage` (allowing `JointState` to override our regular TF graph) is useful in at least two situations:
1. The user is manually adjusting joint positions in Studio and not connected to a live ROS stack (i.e. paused bag playback, or URDF visualized with no other data source loaded)
2. The user has written joint state to a bag/mcap file and for whatever reason hasn't recorded TF messages (forgot to record TF, non-ROS robot, etc).

Per discussion with @jhurliman, overriding TF with joint states should probably be explicitly opt-in.

**Comment by johannespitz** [comment]
Hi, I would love to have exactly that feature. 
We are not using ROS and, of course, I could just compute all the necessary transforms myself, but it looks like all the functionality is already there (setting manual joint angles works) so I assume it wouldn't be too much work to use joint angles directly. 
I'd also be willing to help implementing it, but I have no experience with typescript so I would need some guidance.
